### DISCUSSION TITLE
Multiple Renders With NextJs

### DISCUSSION THREAD

**Discussion by LucasChizzoli** [trpc/trpc#1987] [Q&A] [answered]
Hey all!

I'm using tRCP with NextJs. When using `trcp.useQuery()` in a component I'm getting multiple renders.

Here is my `package.json`

`{
  "name": "nextjs-trpc-poc",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  },
  "dependencies": {
    "@trpc/client": "^9.25.2",
    "@trpc/next": "^9.25.2",
    "@trpc/react": "^9.25.2",
    "@trpc/server": "^9.25.2",
    "next": "12.1.6",
    "react": "18.1.0",
    "react-dom": "18.1.0",
    "react-query": "^3.39.1",
    "zod": "^3.17.3"
  },
  "devDependencies": {
    "@types/node": "17.0.42",
    "@types/react": "18.0.12",
    "@types/react-dom": "18.0.5",
    "eslint": "8.17.0",
    "eslint-config-next": "12.1.6",
    "typescript": "4.7.3"
  }
}
`

And here is a repo were you can see the problem
[https://github.com/LucasChizzoli/nextjs-trpc-poc](https://github.com/LucasChizzoli/nextjs-trpc-poc)

Is this happening to anyone else?

Thanks in advance.

**Comment by jlalmes** [comment]
Haven't tried to run your reproduction, but this sounds like you are describing the result of `reactStrictMode: true` in `react@18`. 

Take a look at this: 
https://stackoverflow.com/questions/48846289/why-is-my-react-component-is-rendering-twice

**Comment by sachinraja** [comment]
tRPC/react-query does a fetch on window mount even with `ssr: true` so that's probably why you're seeing that. I'll spin up an example and link that here.

**Comment by LucasChizzoli** [comment]
> reactStrictMode

I tried but it continues to happen to me. Also according to tRPC documentation `strict` should be `true` for tRPC to work. If I set it to `false` I don't get the types check.

**Comment by jlalmes** [comment]
> reactStrictMode

I meant `reactStrictMode` in your `next.config.js` 😁

**Comment by LucasChizzoli** [comment]
> > reactStrictMode
> 
> I meant `reactStrictMode` in your `next.config.js` 😁

Yes I tried it too. Same behavior. But @sachinraja may be in the correct path. Because I noticed that when navigating away from the browser, for example focusing in the terminal, and going back to the browser, it fires a re-render.

**Comment by sachinraja** [comment, marked as answer]
You can configure `queryClientConfig` to control refetching behavior:
```ts
export default withTRPC<AppRouter>({
  config({ ctx }) {
    /**
     * If you want to use SSR, you need to use the server's full URL
     * @link https://trpc.io/docs/ssr
     */
    const url = process.env.VERCEL_URL
      ? `https://${process.env.VERCEL_URL}/api/trpc`
      : 'http://localhost:3000/api/trpc';

    return {
      url,
      queryClientConfig: {
        defaultOptions: {
          queries: {
            refetchOnMount: false,
            // refetchOnWindowFocus: false
          },
        },
      },
    };
  },
  /**
   * @link https://trpc.io/docs/ssr
   */
  ssr: true,
})(MyApp);
```

**Reply by LucasChizzoli** [reply]
Thank you very much @sachinraja! `refetchOnWindowFocus: false` prevented the refetched!

**Reply by sachinraja** [reply]
FYI this is React Query's default behavior and it is there for a reason.
See
https://tanstack.com/query/v4/docs/guides/important-defaults
https://tanstack.com/query/v4/docs/guides/window-focus-refetching

**Reply by LucasChizzoli** [reply]
Gotcha, I'm new to react-query. In this case it's a super static content so I don't need to refresh it that often. But I'll keep this behavior in mind. Thanks for the help!

**Reply by garrettg123** [reply]
I'm seeing this with that code:
```
Type '({ ctx }: { ctx?: NextPageContext | undefined; }) => { url: string; queryClientConfig: { defaultOptions: { queries: { refetchOnMount: false; refetchOnWindowFocus: false; }; }; }; }' is not assignable to type '(info: { ctx?: NextPageContext | undefined; }) => WithTRPCConfig<CreateRouterInner<RootConfig<{ ctx: { session: Session | null; prisma: PrismaClient<PrismaClientOptions, never, DefaultArgs>; }; meta: object; errorShape: { ...; }; transformer: typeof SuperJSON; }>, { ...; }>>'.
  Type '{ url: string; queryClientConfig: { defaultOptions: { queries: { refetchOnMount: false; refetchOnWindowFocus: false; }; }; }; }' is not assignable to type 'WithTRPCConfig<CreateRouterInner<RootConfig<{ ctx: { session: Session | null; prisma: PrismaClient<PrismaClientOptions, never, DefaultArgs>; }; meta: object; errorShape: { ...; }; transformer: typeof SuperJSON; }>, { ...; }>>'.
    Type '{ url: string; queryClientConfig: { defaultOptions: { queries: { refetchOnMount: false; refetchOnWindowFocus: false; }; }; }; }' is not assignable to type '{ transformer: typeof SuperJSON; } & { links: TRPCLink<CreateRouterInner<RootConfig<{ ctx: { session: Session | null; prisma: PrismaClient<PrismaClientOptions, never, DefaultArgs>; }; meta: object; errorShape: { ...; }; transformer: typeof SuperJSON; }>, { ...; }>>[]; } & { ...; } & { ...; }'.
      Property 'transformer' is missing in type '{ url: string; queryClientConfig: { defaultOptions: { queries: { refetchOnMount: false; refetchOnWindowFocus: false; }; }; }; }' but required in type '{ transformer: typeof SuperJSON; }'.ts(2322)
TRPCUntypedClient.d.ts(20, 5): 'transformer' is declared here.
withTRPC.d.ts(11, 5): The expected type comes from property 'config' which is declared here on type 'WithTRPCNoSSROptions<CreateRouterInner<RootConfig<{ ctx: { session: Session | null; prisma: PrismaClient<PrismaClientOptions, never, DefaultArgs>; }; meta: object; errorShape: { ...; }; transformer: typeof SuperJSON; }>, { ...; }>> | WithTRPCSSROptions<...>'
(property) WithTRPCOptions<CreateRouterInner<RootConfig<{ ctx: { session: Session | null; prisma: PrismaClient<Prisma.PrismaClientOptions, never, DefaultArgs>; }; meta: object; errorShape: { ...; }; transformer: typeof SuperJSON; }>, { ...; }>>.config: (info: {
    ctx?: NextPageContext | undefined;
}) => WithTRPCConfig<CreateRouterInner<RootConfig<{
    ctx: {
        session: Session | null;
        prisma: PrismaClient<Prisma.PrismaClientOptions, never, DefaultArgs>;
    };
    meta: object;
    errorShape: {
        ...;
    };
    transformer: typeof SuperJSON;
}>, {
    ...;
}>>
```

**Reply by garrettg123** [reply]
Fixed by moving this to `api.ts` in `config()`:
```ts
      queryClientConfig: {
        defaultOptions: {
          queries: {
            refetchOnWindowFocus: false,
            refetchOnMount: false,
          },
        },
      },
